pkg/uploader/provider/kopia.go

```diff
@@ -0,0 +1,204 @@
+				kp.log.Infof("Backup is been canceled")
+			}
+			if kp.restoreCancel != nil {
+				close(kp.restoreCancel)
+				kp.log.Infof("Restore is been canceled")
+			}
+			return
+		default:
+			time.Sleep(time.Second * 10)
+		}
+	}
+}
+
+func (kp *kopiaProvider) Close(ctx context.Context) {
+	kp.bkRepo.Close(ctx)
+}
+
+//RunBackup which will backup specific path and update backup progress
+func (kp *kopiaProvider) RunBackup(
+	ctx context.Context,
+	path string,
+	tags map[string]string,
+	parentSnapshot string,
+	updater uploader.ProgressUpdater) (string, error) {
+	if updater == nil {
+		return "", errors.New("Need to initial backup progress updater first")
+	}
+
+	log := kp.log.WithFields(logrus.Fields{
+		"path":           path,
+		"parentSnapshot": parentSnapshot,
+	})
+	repoWriter := kopia.NewShimRepo(kp.bkRepo)
+	kp.uploader = snapshotfs.NewUploader(repoWriter)
+	prorgess := new(kopia.KopiaProgress)
+	prorgess.InitThrottle(backupProgressCheckInterval)
+	prorgess.Updater = updater
+	kp.uploader.Progress = prorgess
+
+	log.Info("Starting backup")
+	go kp.CheckContext(ctx)
+
+	snapshotInfo, err := BackupFunc(ctx, kp.uploader, repoWriter, path, parentSnapshot, log)
+
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to run kopia backup")
+	} else if snapshotInfo == nil {
+		return "", fmt.Errorf("failed to get kopia backup snapshot info for path %v", path)
+	}
+
+	updater.UpdateProgress(
+		&uploader.UploaderProgress{
+			TotalBytes: snapshotInfo.Size,
+			BytesDone:  snapshotInfo.Size,
+		},
+	)
+
+	log.Debugf("Kopia backup finished, snapshot ID %s, backup size %d", snapshotInfo.ID, snapshotInfo.Size)
+	return snapshotInfo.ID, nil
+}
+
+func (kp *kopiaProvider) GetPassword(param interface{}) (string, error) {
+	if kp.credGetter.FromSecret == nil {
+		return "", errors.New("invalid credentials interface")
+	}
+	rawPass, err := kp.credGetter.FromSecret.Get(repokeys.RepoKeySelector())
+	if err != nil {
+		return "", errors.Wrap(err, "error to get password")
+	}
+
+	return strings.TrimSpace(rawPass), nil
+}
+
+//RunRestore which will restore specific path and update restore progress
+func (kp *kopiaProvider) RunRestore(
+	ctx context.Context,
+	snapshotID string,
+	volumePath string,
+	updater uploader.ProgressUpdater) error {
+	log := kp.log.WithFields(logrus.Fields{
+		"snapshotID": snapshotID,
+		"volumePath": volumePath,
+	})
+	repoWriter := kopia.NewShimRepo(kp.bkRepo)
+	prorgess := new(kopia.KopiaProgress)
+	prorgess.InitThrottle(restoreProgressCheckInterval)
+	prorgess.Updater = updater
+	kp.restoreCancel = make(chan struct{})
+	defer func() {
+		if kp.restoreCancel != nil {
+			close(kp.restoreCancel)
+		}
+	}()
+
+	log.Info("Starting restore")
+	go kp.CheckContext(ctx)
+	size, fileCount, err := RestoreFunc(ctx, repoWriter, prorgess, snapshotID, volumePath, log, kp.restoreCancel)
+
+	if err != nil {
+		return errors.Wrapf(err, "Failed to run kopia restore")
+	}
+
+	updater.UpdateProgress(&uploader.UploaderProgress{
```

Comment:
Why do we need to update the progress here? Could you add comments about it if it is needed indeed?

Reply:
To ensure that the statistic data of TotalBytes is equal to BytesDone when finished, I've added one annotation